app/bundles/LeadBundle/Helper/ContactRequestHelper.php

```diff
@@ -0,0 +1,303 @@
+     * @param StatRepository       $emailStatRepository
+     * @param LeadDeviceRepository $leadDeviceRepository
+     * @param RequestStack         $requestStack
+     * @param Logger               $logger
+     */
+    public function __construct(
+        LeadModel $leadModel,
+        ContactTracker $contactTracker,
+        CoreParametersHelper $coreParametersHelper,
+        IpLookupHelper $ipLookupHelper,
+        StatRepository $emailStatRepository,
+        LeadDeviceRepository $leadDeviceRepository,
+        RequestStack $requestStack,
+        Logger $logger
+    ) {
+        $this->leadModel            = $leadModel;
+        $this->contactTracker       = $contactTracker;
+        $this->coreParametersHelper = $coreParametersHelper;
+        $this->ipLookupHelper       = $ipLookupHelper;
+        $this->emailStatRepository  = $emailStatRepository;
+        $this->leadDeviceRepository = $leadDeviceRepository;
+        $this->request              = $requestStack->getCurrentRequest();
+        $this->logger               = $logger;
+    }
+
+    /**
+     * @param Lead  $trackedContact
+     * @param array $queryFields
+     *
+     * @return Lead
+     */
+    public function getContactFromQuery(array $queryFields = [])
+    {
+        $this->trackedContact = $this->contactTracker->getContact();
+        $this->queryFields    = $queryFields;
+
+        if ($foundContact = $this->getContactFromUrl()) {
+            $this->trackedContact = $foundContact;
+            $this->contactTracker->setTrackedContact($this->trackedContact);
+        }
+
+        $this->prepareContactFromRequest();
+
+        return $this->trackedContact;
+    }
+
+    /**
+     * @return Lead|null
+     */
+    private function getContactFromUrl()
+    {
+        // Check for a lead requested through clickthrough query parameter
+        if (isset($this->queryFields['ct'])) {
+            $clickthrough = $this->queryFields['ct'];
+        } elseif ($clickthrough = $this->request->get('ct', [])) {
+            $clickthrough = ClickthroughHelper::decodeArrayFromUrl($clickthrough);
+        }
+
+        if (!is_array($clickthrough)) {
+            return null;
+        }
+
+        if ($contact = $this->getContactFromEmailClickthrough()) {
+            return $contact;
+        }
+
+        $this->setEmailFromClickthroughIdentification();
+
+        /** @var Lead $foundContact */
+        list($foundContact, $this->publiclyUpdatableFieldValues) = $this->leadModel->checkForDuplicateContact(
+            $this->queryFields,
+            $this->trackedContact,
+            true,
+            true
+        );
+        if ($foundContact->getId() !== $this->trackedContact->getId()) {
+            // A contact was found by a publicly updatable field
+            return $foundContact;
+        }
+
+        if ($contact = $this->getContactByFingerprint()) {
+            return $contact;
+        }
+
+        return null;
```

Comment:
We had a discussion about this with @Maxell92 and @galvani yesterday. We agreed it is better to have one return type (Lead) and throw a NotFoundException instead of returning null.

When calling the method it would look like this:
```
try {
    $lead = $this->getContactFromUrl();
    ...
} catch (NotFoundException $e) {
    ...
}
```
instead of
```
$leadOrNull = $this->getContactFromUrl();
if (is_null($leadOrNull) {
    ...
} else {
    ...
}
```
It improves readability where one can clearly see what was the reason for the exception vs no one knows what is the reason for null. And in PHP 7+ we can declare the one and only return type like:
```
private function getContactFromUrl(): Lead {...}
```
It would be a BC break for existing methods (if not private), but for the new ones it's a good idea to start doing that.
What do you think?

Reply:
We can't use PHP 7 yet until we bump minimum version but I'll implement the exception instead of null.

Reply:
That was another reason why to avoid returning multiple return types from one method, but the main message I wanted to point out was to return only one type.